// typeorm/repository/Repository.js
"use strict";
var PlainObjectToNewEntityTransformer_1 = require("../query-builder/transformer/PlainObjectToNewEntityTransformer");
var PlainObjectToDatabaseEntityTransformer_1 = require("../query-builder/transformer/PlainObjectToDatabaseEntityTransformer");
var EntityPersistOperationsBuilder_1 = require("../persistment/EntityPersistOperationsBuilder");
var PersistOperationExecutor_1 = require("../persistment/PersistOperationExecutor");
var FindOptions_1 = require("./FindOptions");
/**
 * Repository is supposed to work with your entity objects. Find entities, insert, update, delete, etc.
 */
var Repository = (function () {
    // -------------------------------------------------------------------------
    // Constructor
    // -------------------------------------------------------------------------
    function Repository(connection, metadata) {
        this.connection = connection;
        this.metadata = metadata;
    }
    // -------------------------------------------------------------------------
    // Public Methods
    // -------------------------------------------------------------------------
    /**
     * Checks if entity has an id.
     */
    Repository.prototype.hasId = function (entity) {
        return entity && this.metadata.primaryColumn && entity.hasOwnProperty(this.metadata.primaryColumn.propertyName);
    };
    /**
     * Creates a new query builder that can be used to build a sql query.
     */
    Repository.prototype.createQueryBuilder = function (alias) {
        return this.connection.driver
            .createQueryBuilder()
            .select(alias)
            .from(this.metadata.target, alias);
    };
    /**
     * Creates a new entity. If fromRawEntity is given then it creates a new entity and copies all entity properties
     * from this object into a new entity (copies only properties that should be in a new entity).
     */
    Repository.prototype.create = function (fromRawEntity) {
        if (fromRawEntity) {
            var transformer = new PlainObjectToNewEntityTransformer_1.PlainObjectToNewEntityTransformer();
            return transformer.transform(fromRawEntity, this.metadata);
        }
        return this.metadata.create();
    };
    /**
     * Creates a entities from the given array of plain javascript objects.
     */
    Repository.prototype.createMany = function (copyFromObjects) {
        var _this = this;
        return copyFromObjects.map(function (object) { return _this.create(object); });
    };
    /**
     * Creates a new entity from the given plan javascript object. If entity already exist in the database, then
     * it loads it (and everything related to it), replaces all values with the new ones from the given object
     * and returns this new entity. This new entity is actually a loaded from the db entity with all properties
     * replaced from the new object.
     */
    Repository.prototype.initialize = function (object) {
        var transformer = new PlainObjectToDatabaseEntityTransformer_1.PlainObjectToDatabaseEntityTransformer();
        var queryBuilder = this.createQueryBuilder(this.metadata.table.name);
        return transformer.transform(object, this.metadata, queryBuilder);
    };
    /**
     * Merges two entities into one new entity.
     */
    Repository.prototype.merge = function (entity1, entity2) {
        return Object.assign(this.metadata.create(), entity1, entity2);
    };
    /**
     * Persists (saves) a given entity in the database. If entity does not exist in the database then it inserts it,
     * else if entity already exist in the database then it updates it.
     */
    Repository.prototype.persist = function (entity) {
        var _this = this;
        var loadedDbEntity;
        var persister = new PersistOperationExecutor_1.PersistOperationExecutor(this.connection);
        var builder = new EntityPersistOperationsBuilder_1.EntityPersistOperationBuilder(this.connection);
        var allPersistedEntities = this.extractObjectsById(entity, this.metadata);
        var promise = !this.hasId(entity) ? Promise.resolve(null) : this.initialize(entity);
        return promise
            .then(function (dbEntity) {
            loadedDbEntity = dbEntity;
            return _this.findNotLoadedIds(_this.extractObjectsById(dbEntity, _this.metadata), allPersistedEntities);
        }) // need to find db entities that were not loaded by initialize method
            .then(function (allDbEntities) {
            var persistOperation = builder.buildFullPersistment(_this.metadata, loadedDbEntity, entity, allDbEntities, allPersistedEntities);
            return persister.executePersistOperation(persistOperation);
        }).then(function () { return entity; });
    };
    /**
     * Removes a given entity from the database.
     */
    Repository.prototype.remove = function (entity) {
        var _this = this;
        var persister = new PersistOperationExecutor_1.PersistOperationExecutor(this.connection);
        return this.initialize(entity).then(function (dbEntity) {
            entity[_this.metadata.primaryColumn.name] = undefined;
            var builder = new EntityPersistOperationsBuilder_1.EntityPersistOperationBuilder(_this.connection);
            var dbEntities = _this.extractObjectsById(dbEntity, _this.metadata);
            var allPersistedEntities = _this.extractObjectsById(entity, _this.metadata);
            var persistOperation = builder.buildOnlyRemovement(_this.metadata, dbEntity, entity, dbEntities, allPersistedEntities);
            return persister.executePersistOperation(persistOperation);
        }).then(function () { return entity; });
    };
    /**
     * Finds entities that match given conditions.
     */
    Repository.prototype.find = function (conditionsOrFindOptions, options) {
        return this.createFindQueryBuilder(conditionsOrFindOptions, options)
            .getResults();
    };
    /**
     * Finds entities that match given conditions.
     */
    Repository.prototype.findAndCount = function (conditionsOrFindOptions, options) {
        var qb = this.createFindQueryBuilder(conditionsOrFindOptions, options);
        return Promise.all([qb.getResults(), qb.getCount()])
            .then(function (_a) {
            var entities = _a[0], count = _a[1];
            return ({ items: entities, count: count });
        });
    };
    /**
     * Finds first entity that matches given conditions.
     */
    Repository.prototype.findOne = function (conditionsOrFindOptions, options) {
        return this.createFindQueryBuilder(conditionsOrFindOptions, options)
            .getSingleResult();
    };
    /**
     * Finds entity with given id.
     */
    Repository.prototype.findById = function (id, options) {
        return this.createFindQueryBuilder((_a = {}, _a[this.metadata.primaryColumn.name] = id, _a), options)
            .getSingleResult();
        var _a;
    };
    /**
     * Executes raw SQL query and returns raw database results.
     */
    Repository.prototype.query = function (query) {
        return this.connection.driver.query(query);
    };
    /**
     * Wraps given function execution (and all operations made there) in a transaction.
     */
    Repository.prototype.transaction = function (runInTransaction) {
        var _this = this;
        var runInTransactionResult;
        return this.connection.driver
            .beginTransaction()
            .then(function () { return runInTransaction(); })
            .then(function (result) {
            runInTransactionResult = result;
            return _this.connection.driver.endTransaction();
        })
            .then(function () { return runInTransactionResult; });
    };
    // -------------------------------------------------------------------------
    // Private Methods
    // -------------------------------------------------------------------------
    Repository.prototype.createFindQueryBuilder = function (conditionsOrFindOptions, options) {
        var findOptions, conditions;
        if (FindOptions_1.FindOptionsUtils.isFindOptions(conditionsOrFindOptions)) {
            findOptions = conditionsOrFindOptions;
        }
        else {
            findOptions = options;
            conditions = conditionsOrFindOptions;
        }
        var alias = findOptions ? options.alias : this.metadata.table.name;
        var qb = this.createQueryBuilder(alias);
        if (findOptions) {
            FindOptions_1.FindOptionsUtils.applyOptionsToQueryBuilder(qb, findOptions);
        }
        if (conditions) {
            Object.keys(conditions).forEach(function (key) {
                var name = key.indexOf(".") === -1 ? alias + "." + key : key;
                qb.andWhere(name + "=:" + key);
            });
            qb.addParameters(conditions);
        }
        return qb;
    };
    /**
     * When ORM loads dbEntity it uses joins to load all entity dependencies. However when dbEntity is newly persisted
     * to the db, but uses already exist in the db relational entities, those entities cannot be loaded, and will
     * absent in dbEntities. To fix it, we need to go throw all persistedEntities we have, find out those which have
     * ids, check if we did not load them yet and try to load them. This algorithm will make sure that all dbEntities
     * are loaded. Further it will help insert operations to work correctly.
     */
    Repository.prototype.findNotLoadedIds = function (dbEntities, persistedEntities) {
        var _this = this;
        var missingDbEntitiesLoad = persistedEntities
            .filter(function (entityWithId) { return entityWithId.id !== null && entityWithId.id !== undefined; })
            .filter(function (entityWithId) { return !dbEntities.find(function (dbEntity) { return dbEntity.entity.constructor === entityWithId.entity.constructor && dbEntity.id === entityWithId.id; }); })
            .map(function (entityWithId) {
            var metadata = _this.connection.getEntityMetadata(entityWithId.entity.constructor);
            var repository = _this.connection.getRepository(entityWithId.entity.constructor);
            return repository.findById(entityWithId.id).then(function (loadedEntity) {
                if (!loadedEntity)
                    return undefined;
                return {
                    id: loadedEntity[metadata.primaryColumn.name],
                    entity: loadedEntity
                };
            });
        });
        return Promise.all(missingDbEntitiesLoad).then(function (missingDbEntities) {
            return dbEntities.concat(missingDbEntities.filter(function (dbEntity) { return !!dbEntity; }));
        });
    };
    /**
     * Extracts unique objects from given entity and all its downside relations.
     */
    Repository.prototype.extractObjectsById = function (entity, metadata, entityWithIds) {
        var _this = this;
        if (entityWithIds === void 0) { entityWithIds = []; }
        if (!entity)
            return entityWithIds;
        metadata.relations
            .filter(function (relation) { return !!entity[relation.propertyName]; })
            .forEach(function (relation) {
            var relMetadata = relation.relatedEntityMetadata;
            var value = entity[relation.propertyName];
            if (value instanceof Array) {
                value.forEach(function (subEntity) { return _this.extractObjectsById(subEntity, relMetadata, entityWithIds); });
            }
            else {
                _this.extractObjectsById(value, relMetadata, entityWithIds);
            }
        });
        if (!entityWithIds.find(function (entityWithId) { return entityWithId.entity === entity; })) {
            entityWithIds.push({
                id: entity[metadata.primaryColumn.name],
                entity: entity
            });
        }
        return entityWithIds;
    };
    return Repository;
}());
exports.Repository = Repository;
