// typeorm/query-builder/QueryBuilder.js
"use strict";
var Alias_1 = require("./alias/Alias");
var AliasMap_1 = require("./alias/AliasMap");
var RawSqlResultsToEntityTransformer_1 = require("./transformer/RawSqlResultsToEntityTransformer");
var Broadcaster_1 = require("../subscriber/Broadcaster");
var QueryBuilder = (function () {
    // -------------------------------------------------------------------------
    // Constructor
    // -------------------------------------------------------------------------
    function QueryBuilder(connection) {
        this.connection = connection;
        this.selects = [];
        this.joins = [];
        this.groupBys = [];
        this.wheres = [];
        this.havings = [];
        this.orderBys = [];
        this.parameters = {};
        this._aliasMap = new AliasMap_1.AliasMap(connection.entityMetadatas);
        this.broadcaster = new Broadcaster_1.Broadcaster(connection);
    }
    Object.defineProperty(QueryBuilder.prototype, "aliasMap", {
        // -------------------------------------------------------------------------
        // Accessors
        // -------------------------------------------------------------------------
        get: function () {
            return this._aliasMap;
        },
        enumerable: true,
        configurable: true
    });
    QueryBuilder.prototype.delete = function (tableNameOrEntity) {
        if (tableNameOrEntity instanceof Function) {
            var aliasName = tableNameOrEntity.name;
            var aliasObj = new Alias_1.Alias(aliasName);
            aliasObj.target = tableNameOrEntity;
            this._aliasMap.addMainAlias(aliasObj);
            this.fromEntity = { alias: aliasObj };
        }
        else if (typeof tableNameOrEntity === "string") {
            this.fromTableName = tableNameOrEntity;
        }
        this.type = "delete";
        return this;
    };
    QueryBuilder.prototype.update = function (tableNameOrEntityOrUpdateSet, updateSet) {
        if (tableNameOrEntityOrUpdateSet instanceof Function) {
            var aliasName = tableNameOrEntityOrUpdateSet.name;
            var aliasObj = new Alias_1.Alias(aliasName);
            aliasObj.target = tableNameOrEntityOrUpdateSet;
            this._aliasMap.addMainAlias(aliasObj);
            this.fromEntity = { alias: aliasObj };
        }
        else if (typeof tableNameOrEntityOrUpdateSet === "object") {
            updateSet = tableNameOrEntityOrUpdateSet;
        }
        else if (typeof tableNameOrEntityOrUpdateSet === "string") {
            this.fromTableName = tableNameOrEntityOrUpdateSet;
        }
        this.type = "update";
        this.updateQuerySet = updateSet;
        return this;
    };
    QueryBuilder.prototype.select = function (selection) {
        this.type = "select";
        if (selection) {
            if (selection instanceof Array) {
                this.selects = selection;
            }
            else {
                this.selects = [selection];
            }
        }
        return this;
    };
    QueryBuilder.prototype.addSelect = function (selection) {
        if (selection instanceof Array)
            this.selects = this.selects.concat(selection);
        else
            this.selects.push(selection);
        return this;
    };
    QueryBuilder.prototype.from = function (entityOrTableName, alias) {
        if (entityOrTableName instanceof Function) {
            var aliasObj = new Alias_1.Alias(alias);
            aliasObj.target = entityOrTableName;
            this._aliasMap.addMainAlias(aliasObj);
            this.fromEntity = { alias: aliasObj };
        }
        else {
            this.fromTableName = entityOrTableName;
            this.fromTableAlias = alias;
        }
        return this;
    };
    QueryBuilder.prototype.innerJoinAndSelect = function (entityOrProperty, alias, conditionType, condition) {
        this.addSelect(alias);
        return this.join("inner", entityOrProperty, alias, conditionType, condition);
    };
    QueryBuilder.prototype.innerJoin = function (entityOrProperty, alias, conditionType, condition) {
        return this.join("inner", entityOrProperty, alias, conditionType, condition);
    };
    QueryBuilder.prototype.leftJoinAndSelect = function (entityOrProperty, alias, conditionType, condition) {
        if (conditionType === void 0) { conditionType = "on"; }
        this.addSelect(alias);
        return this.join("left", entityOrProperty, alias, conditionType, condition);
    };
    QueryBuilder.prototype.leftJoin = function (entityOrProperty, alias, conditionType, condition) {
        if (conditionType === void 0) { conditionType = "on"; }
        return this.join("left", entityOrProperty, alias, conditionType, condition);
    };
    QueryBuilder.prototype.join = function (joinType, entityOrProperty, alias, conditionType, condition) {
        if (conditionType === void 0) { conditionType = "on"; }
        var aliasObj = new Alias_1.Alias(alias);
        this._aliasMap.addAlias(aliasObj);
        if (entityOrProperty instanceof Function) {
            aliasObj.target = entityOrProperty;
        }
        else if (typeof entityOrProperty === "string") {
            aliasObj.parentAliasName = entityOrProperty.split(".")[0];
            aliasObj.parentPropertyName = entityOrProperty.split(".")[1];
        }
        var join = { type: joinType, alias: aliasObj, conditionType: conditionType, condition: condition };
        this.joins.push(join);
        return this;
    };
    QueryBuilder.prototype.where = function (where, parameters) {
        this.wheres.push({ type: "simple", condition: where });
        if (parameters)
            this.addParameters(parameters);
        return this;
    };
    QueryBuilder.prototype.andWhere = function (where, parameters) {
        this.wheres.push({ type: "and", condition: where });
        if (parameters)
            this.addParameters(parameters);
        return this;
    };
    QueryBuilder.prototype.orWhere = function (where, parameters) {
        this.wheres.push({ type: "or", condition: where });
        if (parameters)
            this.addParameters(parameters);
        return this;
    };
    QueryBuilder.prototype.groupBy = function (groupBy) {
        this.groupBys = [groupBy];
        return this;
    };
    QueryBuilder.prototype.addGroupBy = function (groupBy) {
        this.groupBys.push(groupBy);
        return this;
    };
    QueryBuilder.prototype.having = function (having, parameters) {
        this.havings.push({ type: "simple", condition: having });
        if (parameters)
            this.addParameters(parameters);
        return this;
    };
    QueryBuilder.prototype.andHaving = function (having, parameters) {
        this.havings.push({ type: "and", condition: having });
        if (parameters)
            this.addParameters(parameters);
        return this;
    };
    QueryBuilder.prototype.orHaving = function (having, parameters) {
        this.havings.push({ type: "or", condition: having });
        if (parameters)
            this.addParameters(parameters);
        return this;
    };
    QueryBuilder.prototype.orderBy = function (sort, order) {
        if (order === void 0) { order = "ASC"; }
        this.orderBys = [{ sort: sort, order: order }];
        return this;
    };
    QueryBuilder.prototype.addOrderBy = function (sort, order) {
        if (order === void 0) { order = "ASC"; }
        this.orderBys.push({ sort: sort, order: order });
        return this;
    };
    QueryBuilder.prototype.setLimit = function (limit) {
        this.limit = limit;
        return this;
    };
    QueryBuilder.prototype.setOffset = function (offset) {
        this.offset = offset;
        return this;
    };
    QueryBuilder.prototype.setFirstResult = function (firstResult) {
        this.firstResult = firstResult;
        return this;
    };
    QueryBuilder.prototype.setMaxResults = function (maxResults) {
        this.maxResults = maxResults;
        return this;
    };
    QueryBuilder.prototype.setParameter = function (key, value) {
        this.parameters[key] = value;
        return this;
    };
    QueryBuilder.prototype.setParameters = function (parameters) {
        var _this = this;
        this.parameters = {};
        Object.keys(parameters).forEach(function (key) { return _this.parameters[key] = parameters[key]; });
        return this;
    };
    QueryBuilder.prototype.addParameters = function (parameters) {
        var _this = this;
        Object.keys(parameters).forEach(function (key) { return _this.parameters[key] = parameters[key]; });
        return this;
    };
    QueryBuilder.prototype.getSql = function () {
        var sql = this.createSelectExpression();
        sql += this.createJoinExpression();
        sql += this.createWhereExpression();
        sql += this.createGroupByExpression();
        sql += this.createHavingExpression();
        sql += this.createOrderByExpression();
        sql += this.createLimitExpression();
        sql += this.createOffsetExpression();
        sql = this.replaceParameters(sql);
        return sql;
    };
    QueryBuilder.prototype.execute = function () {
        return this.connection.driver.query(this.getSql());
    };
    QueryBuilder.prototype.getScalarResults = function () {
        return this.connection.driver.query(this.getSql());
    };
    QueryBuilder.prototype.getSingleScalarResult = function () {
        return this.getScalarResults().then(function (results) { return results[0]; });
    };
    QueryBuilder.prototype.getResults = function () {
        var _this = this;
        var mainAlias = this.aliasMap.mainAlias.name;
        if (this.firstResult || this.maxResults) {
            var metadata_1 = this.connection.getEntityMetadata(this.fromEntity.alias.target);
            var idsQuery = this.clone()
                .select("DISTINCT(" + mainAlias + "." + metadata_1.primaryColumn.name + ") as ids")
                .setOffset(this.firstResult)
                .setLimit(this.maxResults)
                .getSql();
            return this.connection.driver
                .query(idsQuery)
                .then(function (results) {
                var ids = results.map(function (result) { return result["ids"]; }).join(", ");
                var queryWithIds = _this.clone()
                    .andWhere(mainAlias + "." + metadata_1.primaryColumn.name + " IN (" + ids + ")")
                    .getSql();
                return _this.connection.driver.query(queryWithIds);
            })
                .then(function (results) { return _this.rawResultsToEntities(results); })
                .then(function (results) { return _this.broadcaster.broadcastLoadEventsForAll(results).then(function () { return results; }); });
        }
        else {
            return this.connection.driver
                .query(this.getSql())
                .then(function (results) { return _this.rawResultsToEntities(results); })
                .then(function (results) { return _this.broadcaster.broadcastLoadEventsForAll(results).then(function () { return results; }); });
        }
    };
    QueryBuilder.prototype.getSingleResult = function () {
        return this.getResults().then(function (entities) { return entities[0]; });
    };
    QueryBuilder.prototype.getCount = function () {
        var mainAlias = this.aliasMap.mainAlias.name;
        var metadata = this.connection.getEntityMetadata(this.fromEntity.alias.target);
        var countQuery = this.clone()
            .select("COUNT(DISTINCT(" + mainAlias + "." + metadata.primaryColumn.name + ")) as cnt")
            .getSql();
        return this.connection.driver
            .query(countQuery)
            .then(function (results) { return parseInt(results[0]["cnt"]); });
    };
    QueryBuilder.prototype.clone = function () {
        var _this = this;
        var qb = new QueryBuilder(this.connection);
        switch (this.type) {
            case "select":
                qb.select(this.selects);
                break;
            case "update":
                qb.update(this.updateQuerySet);
                break;
            case "delete":
                qb.delete();
                break;
        }
        if (this.fromEntity && this.fromEntity.alias && this.fromEntity.alias.target) {
            qb.from(this.fromEntity.alias.target, this.fromEntity.alias.name);
        }
        else if (this.fromTableName) {
            qb.from(this.fromTableName, this.fromTableAlias);
        }
        this.joins.forEach(function (join) {
            var property = join.alias.target || (join.alias.parentAliasName + "." + join.alias.parentPropertyName);
            qb.join(join.type, property, join.alias.name, join.conditionType, join.condition);
        });
        this.groupBys.forEach(function (groupBy) { return qb.addGroupBy(groupBy); });
        this.wheres.forEach(function (where) {
            switch (where.type) {
                case "simple":
                    qb.where(where.condition);
                    break;
                case "and":
                    qb.andWhere(where.condition);
                    break;
                case "or":
                    qb.orWhere(where.condition);
                    break;
            }
        });
        this.havings.forEach(function (having) {
            switch (having.type) {
                case "simple":
                    qb.where(having.condition);
                    break;
                case "and":
                    qb.andWhere(having.condition);
                    break;
                case "or":
                    qb.orWhere(having.condition);
                    break;
            }
        });
        this.orderBys.forEach(function (orderBy) { return qb.addOrderBy(orderBy.sort, orderBy.order); });
        Object.keys(this.parameters).forEach(function (key) { return qb.setParameter(key, _this.parameters[key]); });
        qb.setLimit(this.limit)
            .setOffset(this.offset)
            .setFirstResult(this.firstResult)
            .setMaxResults(this.maxResults);
        return qb;
    };
    // -------------------------------------------------------------------------
    // Protected Methods
    // -------------------------------------------------------------------------
    QueryBuilder.prototype.rawResultsToEntities = function (results) {
        var transformer = new RawSqlResultsToEntityTransformer_1.RawSqlResultsToEntityTransformer(this._aliasMap);
        return transformer.transform(results);
    };
    QueryBuilder.prototype.createSelectExpression = function () {
        // todo throw exception if selects or from is missing
        var _this = this;
        var alias, tableName;
        var allSelects = [];
        if (this.fromEntity) {
            var metadata = this._aliasMap.getEntityMetadataByAlias(this.fromEntity.alias);
            tableName = metadata.table.name;
            alias = this.fromEntity.alias.name;
            // add select from the main table
            if (this.selects.indexOf(alias) !== -1) {
                metadata.columns.forEach(function (column) {
                    allSelects.push(alias + "." + column.name + " AS " + alias + "_" + column.name);
                });
            }
        }
        else if (this.fromTableName) {
            tableName = this.fromTableName;
        }
        else {
            throw new Error("No from given");
        }
        // add selects from joins
        this.joins
            .filter(function (join) { return _this.selects.indexOf(join.alias.name) !== -1; })
            .forEach(function (join) {
            var joinMetadata = _this._aliasMap.getEntityMetadataByAlias(join.alias);
            joinMetadata.columns.forEach(function (column) {
                allSelects.push(join.alias.name + "." + column.name + " AS " + join.alias.name + "_" + column.propertyName);
            });
        });
        // add all other selects
        this.selects.filter(function (select) {
            return select !== alias && !_this.joins.find(function (join) { return join.alias.name === select; });
        }).forEach(function (select) { return allSelects.push(select); });
        // if still selection is empty, then simply set it to all (*)
        if (allSelects.length === 0)
            allSelects.push("*");
        // create a selection query
        switch (this.type) {
            case "select":
                return "SELECT " + allSelects.join(", ") + " FROM " + tableName + " " + alias;
            case "delete":
                return "DELETE FROM " + tableName + " " + (alias ? alias : "");
            case "update":
                var updateSet = Object.keys(this.updateQuerySet).map(function (key) { return key + "=:updateQuerySet_" + key; });
                var params = Object.keys(this.updateQuerySet).reduce(function (object, key) {
                    object["updateQuerySet_" + key] = _this.updateQuerySet[key];
                    return object;
                }, {});
                this.addParameters(params);
                return "UPDATE " + tableName + " " + (alias ? alias : "") + " SET " + updateSet;
        }
        throw new Error("No query builder type is specified.");
    };
    QueryBuilder.prototype.createWhereExpression = function () {
        if (!this.wheres || !this.wheres.length)
            return "";
        return " WHERE " + this.wheres.map(function (where, index) {
            switch (where.type) {
                case "and":
                    return (index > 0 ? "AND " : "") + where.condition;
                case "or":
                    return (index > 0 ? "OR " : "") + where.condition;
                default:
                    return where.condition;
            }
        }).join(" ");
    };
    QueryBuilder.prototype.createJoinExpression = function () {
        var _this = this;
        return this.joins.map(function (join) {
            var joinType = join.type === "inner" ? "INNER" : "LEFT";
            var appendedCondition = join.condition ? " AND " + join.condition : "";
            var parentAlias = join.alias.parentAliasName;
            var parentMetadata = _this._aliasMap.getEntityMetadataByAlias(_this._aliasMap.findAliasByName(parentAlias));
            var parentTable = parentMetadata.table.name;
            var parentTableColumn = parentMetadata.primaryColumn.name;
            var relation = parentMetadata.findRelationWithDbName(join.alias.parentPropertyName);
            var junctionMetadata = relation.junctionEntityMetadata;
            var joinMetadata = _this._aliasMap.getEntityMetadataByAlias(join.alias);
            var joinTable = joinMetadata.table.name;
            var joinTableColumn = joinMetadata.primaryColumn.name;
            if (relation.isManyToMany) {
                var junctionTable = junctionMetadata.table.name;
                var junctionAlias = join.alias.parentAliasName + "_" + join.alias.name;
                var joinAlias = join.alias.name;
                var condition1 = junctionAlias + "." + junctionMetadata.columns[0].name + "=" + parentAlias + "." + joinTableColumn; // todo: use column names from junction table somehow
                var condition2 = joinAlias + "." + joinTableColumn + "=" + junctionAlias + "." + junctionMetadata.columns[1].name;
                return " " + joinType + " JOIN " + junctionTable + " " + junctionAlias + " " + join.conditionType + " " + condition1 +
                    " " + joinType + " JOIN " + joinTable + " " + joinAlias + " " + join.conditionType + " " + condition2 + appendedCondition;
            }
            else if (relation.isManyToOne || (relation.isOneToOne && relation.isOwning)) {
                var condition = join.alias.name + "." + joinTableColumn + "=" + parentAlias + "." + join.alias.parentPropertyName;
                return " " + joinType + " JOIN " + joinTable + " " + join.alias.name + " " + join.conditionType + " " + condition + appendedCondition;
            }
            else if (relation.isOneToMany || (relation.isOneToOne && !relation.isOwning)) {
                var condition = join.alias.name + "." + relation.inverseSideProperty + "=" + parentAlias + "." + joinTableColumn;
                return " " + joinType + " JOIN " + joinTable + " " + join.alias.name + " " + join.conditionType + " " + condition + appendedCondition;
            }
            else {
                return " " + joinType + " JOIN " + joinTable + " " + join.alias.name + " " + join.conditionType + " " + join.condition;
            }
        }).join(" ");
    };
    QueryBuilder.prototype.createGroupByExpression = function () {
        if (!this.groupBys || !this.groupBys.length)
            return "";
        return " GROUP BY " + this.groupBys.join(", ");
    };
    QueryBuilder.prototype.createHavingExpression = function () {
        if (!this.havings || !this.havings.length)
            return "";
        return " HAVING " + this.havings.map(function (having) {
            switch (having.type) {
                case "and":
                    return " AND " + having.condition;
                case "or":
                    return " OR " + having.condition;
                default:
                    return " " + having.condition;
            }
        }).join(" ");
    };
    QueryBuilder.prototype.createOrderByExpression = function () {
        if (!this.orderBys || !this.orderBys.length)
            return "";
        return " ORDER BY " + this.orderBys.map(function (order) { return order.sort + " " + order.order; }).join(", ");
    };
    QueryBuilder.prototype.createLimitExpression = function () {
        if (!this.limit)
            return "";
        return " LIMIT " + this.limit;
    };
    QueryBuilder.prototype.createOffsetExpression = function () {
        if (!this.offset)
            return "";
        return " OFFSET " + this.offset;
    };
    QueryBuilder.prototype.replaceParameters = function (sql) {
        var _this = this;
        // todo: proper escape values and prevent sql injection
        Object.keys(this.parameters).forEach(function (key) {
            var value = _this.parameters[key] !== null && _this.parameters[key] !== undefined ? "\"" + _this.parameters[key] + "\"" : "NULL";
            sql = sql.replace(":" + key, value); // .replace('"', '')
        });
        return sql;
    };
    return QueryBuilder;
}());
exports.QueryBuilder = QueryBuilder;
