// typeorm/connection/Connection.js
"use strict";
var Repository_1 = require("../repository/Repository");
var RepositoryNotFoundError_1 = require("./error/RepositoryNotFoundError");
var SchemaCreator_1 = require("../schema-creator/SchemaCreator");
var MetadataNotFoundError_1 = require("./error/MetadataNotFoundError");
var EntityManager_1 = require("../repository/EntityManager");
/**
 * A single connection instance to the database. Each connection has its own repositories, subscribers and metadatas.
 */
var Connection = (function () {
    // -------------------------------------------------------------------------
    // Constructor
    // -------------------------------------------------------------------------
    function Connection(name, driver, options) {
        this._entityMetadatas = [];
        this._entityListenerMetadatas = [];
        this._subscribers = [];
        this.repositoryAndMetadatas = [];
        this._name = name;
        this._driver = driver;
        this._driver.connection = this;
        this._options = options;
        this.entityManager = new EntityManager_1.EntityManager(this);
    }
    Object.defineProperty(Connection.prototype, "name", {
        // -------------------------------------------------------------------------
        // Getter / Setter Methods
        // -------------------------------------------------------------------------
        /**
         * The name of the connection.
         */
        get: function () {
            return this._name;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Connection.prototype, "driver", {
        /**
         * Database driver used by this connection.
         */
        get: function () {
            return this._driver;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Connection.prototype, "subscribers", {
        /**
         * All subscribers that are registered for this connection.
         */
        get: function () {
            return this._subscribers;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Connection.prototype, "entityMetadatas", {
        /**
         * All entity metadatas that are registered for this connection.
         */
        get: function () {
            return this._entityMetadatas;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Connection.prototype, "entityListeners", {
        /**
         * All entity listener metadatas that are registered for this connection.
         */
        get: function () {
            return this._entityListenerMetadatas;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Connection.prototype, "repositories", {
        /**
         * All repositories that are registered for this connection.
         */
        get: function () {
            return this.repositoryAndMetadatas.map(function (repoAndMeta) { return repoAndMeta.repository; });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Connection.prototype, "options", {
        /**
         * This connection options and settings.
         */
        get: function () {
            return this._options;
        },
        enumerable: true,
        configurable: true
    });
    // -------------------------------------------------------------------------
    // Public Methods
    // -------------------------------------------------------------------------
    /**
     * Performs connection to the database.
     */
    Connection.prototype.connect = function () {
        var _this = this;
        var schemaCreator = new SchemaCreator_1.SchemaCreator(this);
        return this._driver.connect().then(function () {
            if (_this._options.autoSchemaCreate === true)
                return schemaCreator.create();
        });
    };
    /**
     * Closes this connection.
     */
    Connection.prototype.close = function () {
        return this._driver.disconnect();
    };
    /**
     * Gets EntityManager of this connection.
     */
    Connection.prototype.getEntityManager = function () {
        return this.entityManager;
    };
    /**
     * Gets repository for the given entity class.
     */
    Connection.prototype.getRepository = function (entityClass) {
        var metadata = this.getEntityMetadata(entityClass);
        var repoMeta = this.repositoryAndMetadatas.find(function (repoMeta) { return repoMeta.metadata === metadata; });
        if (!repoMeta)
            throw new RepositoryNotFoundError_1.RepositoryNotFoundError(entityClass);
        return repoMeta.repository;
    };
    /**
     * Gets the entity metadata for the given entity class.
     */
    Connection.prototype.getEntityMetadata = function (entityClass) {
        var metadata = this.entityMetadatas.find(function (metadata) { return metadata.target === entityClass; });
        if (!metadata)
            throw new MetadataNotFoundError_1.MetadataNotFoundError(entityClass);
        return metadata;
    };
    /**
     * Registers entity metadatas for the current connection.
     */
    Connection.prototype.addEntityMetadatas = function (metadatas) {
        var _this = this;
        this._entityMetadatas = this._entityMetadatas.concat(metadatas);
        this.repositoryAndMetadatas = this.repositoryAndMetadatas.concat(metadatas.map(function (metadata) { return _this.createRepoMeta(metadata); }));
        return this;
    };
    /**
     * Registers entity listener metadatas for the current connection.
     */
    Connection.prototype.addEntityListenerMetadatas = function (metadatas) {
        this._entityListenerMetadatas = this._entityListenerMetadatas.concat(metadatas);
        return this;
    };
    /**
     * Registers subscribers for the current connection.
     */
    Connection.prototype.addSubscribers = function (subscribers) {
        this._subscribers = this._subscribers.concat(subscribers);
        return this;
    };
    // -------------------------------------------------------------------------
    // Private Methods
    // -------------------------------------------------------------------------
    Connection.prototype.createRepoMeta = function (metadata) {
        return {
            metadata: metadata,
            repository: new Repository_1.Repository(this, metadata)
        };
    };
    return Connection;
}());
exports.Connection = Connection;
