// typeorm/connection/ConnectionManager.js
"use strict";
var MetadataStorage_1 = require("../metadata-builder/MetadataStorage");
var DefaultNamingStrategy_1 = require("../naming-strategy/DefaultNamingStrategy");
var ConnectionNotFoundError_1 = require("./error/ConnectionNotFoundError");
var EntityMetadataBuilder_1 = require("../metadata-builder/EntityMetadataBuilder");
var DirectoryExportedClassesLoader_1 = require("../util/DirectoryExportedClassesLoader");
var CannotSetNamingStrategyError_1 = require("./error/CannotSetNamingStrategyError");
var ConnectionArray_1 = require("./ConnectionArray");
/**
 * Connection manager holds all connections made to the databases and providers helper management functions
 * for all exist connections.
 */
var ConnectionManager = (function () {
    function ConnectionManager() {
        // -------------------------------------------------------------------------
        // Properties
        // -------------------------------------------------------------------------
        this.connections = new ConnectionArray_1.ConnectionArray();
    }
    Object.defineProperty(ConnectionManager.prototype, "container", {
        // -------------------------------------------------------------------------
        // Accessors
        // -------------------------------------------------------------------------
        /**
         * Sets a container that can be used in custom user subscribers. This allows to inject services in subscribers.
         */
        set: function (container) {
            this._container = container;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ConnectionManager.prototype, "namingStrategy", {
        /**
         * Sets the naming strategy to be used instead of DefaultNamingStrategy.
         */
        set: function (namingStrategy) {
            // if entity metadata builder already initialized then strategy already is used there, and setting a new naming
            // strategy is pointless
            if (this._entityMetadataBuilder)
                throw new CannotSetNamingStrategyError_1.CannotSetNamingStrategyError();
            this._namingStrategy = namingStrategy;
        },
        enumerable: true,
        configurable: true
    });
    ConnectionManager.prototype.createConnection = function (nameOrDriver, driver, options) {
        if (typeof nameOrDriver === "object") {
            driver = nameOrDriver;
            options = driver;
        }
        var name = typeof nameOrDriver === "string" ? nameOrDriver : "default";
        return this.connections.createAndPush(name, driver, options);
    };
    /**
     * Gets registered connection with the given name. If connection name is not given then it will get a default
     * connection.
     */
    ConnectionManager.prototype.getConnection = function (name) {
        if (name === void 0) { name = "default"; }
        if (!name)
            name = "default";
        if (!this.connections.hasWithName(name))
            throw new ConnectionNotFoundError_1.ConnectionNotFoundError(name);
        return this.connections.getWithName(name);
    };
    ConnectionManager.prototype.importEntitiesFromDirectories = function (connectionNameOrPaths, paths) {
        var connectionName = "default";
        if (paths) {
            connectionName = connectionNameOrPaths;
        }
        else {
            paths = connectionNameOrPaths;
        }
        this.importEntities(connectionName, DirectoryExportedClassesLoader_1.importClassesFromDirectories(paths));
    };
    ConnectionManager.prototype.importSubscribersFromDirectories = function (connectionNameOrPaths, paths) {
        var connectionName = "default";
        if (paths) {
            connectionName = connectionNameOrPaths;
        }
        else {
            paths = connectionNameOrPaths;
        }
        this.importSubscribers(connectionName, DirectoryExportedClassesLoader_1.importClassesFromDirectories(paths));
    };
    ConnectionManager.prototype.importEntities = function (connectionNameOrEntities, entities) {
        var connectionName = "default";
        if (entities) {
            connectionName = connectionNameOrEntities;
        }
        else {
            entities = connectionNameOrEntities;
        }
        var entityMetadatas = this.entityMetadataBuilder.build(entities);
        var entityListenerMetadatas = MetadataStorage_1.defaultMetadataStorage.findEntityListenersForClasses(entities);
        this.getConnection(connectionName)
            .addEntityMetadatas(entityMetadatas)
            .addEntityListenerMetadatas(entityListenerMetadatas);
    };
    ConnectionManager.prototype.importSubscribers = function (connectionNameOrSubscriberClasses, subscriberClasses) {
        var _this = this;
        var connectionName = "default";
        if (subscriberClasses) {
            connectionName = connectionNameOrSubscriberClasses;
        }
        else {
            subscriberClasses = connectionNameOrSubscriberClasses;
        }
        var subscribers = MetadataStorage_1.defaultMetadataStorage
            .findEventSubscribersForClasses(subscriberClasses)
            .map(function (metadata) { return _this.createContainerInstance(metadata.target); });
        this.getConnection(connectionName).addSubscribers(subscribers);
    };
    Object.defineProperty(ConnectionManager.prototype, "entityMetadataBuilder", {
        // -------------------------------------------------------------------------
        // Private Methods
        // -------------------------------------------------------------------------
        /**
         * We need to lazily initialize EntityMetadataBuilder because naming strategy can be set before we use entityMetadataBuilder.
         */
        get: function () {
            if (!this._entityMetadataBuilder) {
                var namingStrategy = this._namingStrategy ? this._namingStrategy : new DefaultNamingStrategy_1.DefaultNamingStrategy();
                this._entityMetadataBuilder = new EntityMetadataBuilder_1.EntityMetadataBuilder(MetadataStorage_1.defaultMetadataStorage, namingStrategy);
            }
            return this._entityMetadataBuilder;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Creates a new instance of the given constructor.
     */
    ConnectionManager.prototype.createContainerInstance = function (constructor) {
        return this._container ? this._container.get(constructor) : new constructor();
    };
    return ConnectionManager;
}());
exports.ConnectionManager = ConnectionManager;
