// typeorm/decorator/columns/Column.js
"use strict";
var ColumnTypeUndefinedError_1 = require("../error/ColumnTypeUndefinedError");
var AutoIncrementOnlyForPrimaryError_1 = require("../error/AutoIncrementOnlyForPrimaryError");
var MetadataStorage_1 = require("../../metadata-builder/MetadataStorage");
var ColumnMetadata_1 = require("../../metadata-builder/metadata/ColumnMetadata");
var ColumnTypes_1 = require("../../metadata-builder/types/ColumnTypes");
require("reflect-metadata");
/**
 * Column decorator is used to mark a specific class property as a table column. Only properties decorated with this
 * decorator will be persisted to the database when entity be saved.
 */
function Column(typeOrOptions, options) {
    var type;
    if (typeof typeOrOptions === "string") {
        type = typeOrOptions;
    }
    else {
        options = typeOrOptions;
    }
    return function (object, propertyName) {
        // if type is not given implicitly then try to guess it
        if (!type)
            type = ColumnTypes_1.ColumnTypes.determineTypeFromFunction(Reflect.getMetadata("design:type", object, propertyName));
        // if column options are not given then create a new empty options
        if (!options)
            options = {};
        // check if there is no type in column options then set type from first function argument, or guessed one
        if (!options.type)
            options.type = type;
        // if we still don't have a type then we need to give error to user that type is required
        if (!options.type)
            throw new ColumnTypeUndefinedError_1.ColumnTypeUndefinedError(object, propertyName);
        // check if auto increment is not set for simple column
        if (options.autoIncrement)
            throw new AutoIncrementOnlyForPrimaryError_1.AutoIncrementOnlyForPrimaryError(object, propertyName);
        // create and register a new column metadata
        MetadataStorage_1.defaultMetadataStorage.addColumnMetadata(new ColumnMetadata_1.ColumnMetadata({
            target: object.constructor,
            propertyName: propertyName,
            options: options
        }));
    };
}
exports.Column = Column;
