// typeorm/decorator/columns/PrimaryColumn.js
"use strict";
var ColumnTypes_1 = require("../../metadata-builder/types/ColumnTypes");
var ColumnTypeUndefinedError_1 = require("../error/ColumnTypeUndefinedError");
var MetadataStorage_1 = require("../../metadata-builder/MetadataStorage");
var ColumnMetadata_1 = require("../../metadata-builder/metadata/ColumnMetadata");
var PrimaryColumnCannotBeNullableError_1 = require("../error/PrimaryColumnCannotBeNullableError");
require("reflect-metadata");
/**
 * Column decorator is used to mark a specific class property as a table column. Only properties decorated with this
 * decorator will be persisted to the database when entity be saved. Primary columns also creates a PRIMARY KEY for
 * this column in a db.
 */
function PrimaryColumn(typeOrOptions, options) {
    var type;
    if (typeof typeOrOptions === "string") {
        type = typeOrOptions;
    }
    else {
        options = typeOrOptions;
    }
    return function (object, propertyName) {
        // if type is not given implicitly then try to guess it
        if (!type)
            type = ColumnTypes_1.ColumnTypes.determineTypeFromFunction(Reflect.getMetadata("design:type", object, propertyName));
        // if column options are not given then create a new empty options
        if (!options)
            options = {};
        // check if there is no type in column options then set type from first function argument, or guessed one
        if (!options.type)
            options.type = type;
        // if we still don't have a type then we need to give error to user that type is required
        if (!options.type)
            throw new ColumnTypeUndefinedError_1.ColumnTypeUndefinedError(object, propertyName);
        // check if column is not nullable, because we cannot allow a primary key to be nullable
        if (options.nullable)
            throw new PrimaryColumnCannotBeNullableError_1.PrimaryColumnCannotBeNullableError(object, propertyName);
        // create and register a new column metadata
        MetadataStorage_1.defaultMetadataStorage.addColumnMetadata(new ColumnMetadata_1.ColumnMetadata({
            target: object.constructor,
            propertyName: propertyName,
            isPrimaryKey: true,
            options: options
        }));
    };
}
exports.PrimaryColumn = PrimaryColumn;
