// typeorm/subscriber/Broadcaster.js
"use strict";
var EventListenerTypes_1 = require("../metadata-builder/types/EventListenerTypes");
/**
 * Broadcaster provides a helper methods to broadcast events to the subscribers.
 */
var Broadcaster = (function () {
    // -------------------------------------------------------------------------
    // Constructor
    // -------------------------------------------------------------------------
    function Broadcaster(connection) {
        this.connection = connection;
    }
    // -------------------------------------------------------------------------
    // Accessors
    // -------------------------------------------------------------------------
    Broadcaster.prototype.broadcastBeforeInsertEvent = function (entity) {
        var _this = this;
        var subscribers = this.connection
            .subscribers
            .filter(function (subscriber) { return _this.isAllowedSubscribers(subscriber, entity); })
            .filter(function (subscriber) { return !!subscriber.beforeInsert; })
            .map(function (subscriber) { return subscriber.beforeInsert({ entity: entity }); });
        var listeners = this.connection.entityListeners
            .filter(function (listener) { return listener.type === EventListenerTypes_1.EventListenerTypes.BEFORE_INSERT; })
            .filter(function (listener) { return listener.target === entity.constructor; })
            .map(function (entityListener) { return entity[entityListener.propertyName](); });
        return Promise.all(subscribers.concat(listeners)).then(function () { });
    };
    Broadcaster.prototype.broadcastBeforeUpdateEvent = function (entity, updatedColumns) {
        var _this = this;
        var subscribers = this.connection
            .subscribers
            .filter(function (subscriber) { return _this.isAllowedSubscribers(subscriber, entity); })
            .filter(function (subscriber) { return !!subscriber.beforeUpdate; })
            .map(function (subscriber) { return subscriber.beforeUpdate({ entity: entity, updatedColumns: updatedColumns }); });
        var listeners = this.connection.entityListeners
            .filter(function (listener) { return listener.type === EventListenerTypes_1.EventListenerTypes.BEFORE_UPDATE; })
            .filter(function (listener) { return listener.target === entity.constructor; })
            .map(function (entityListener) { return entity[entityListener.propertyName](); });
        return Promise.all(subscribers.concat(listeners)).then(function () { });
    };
    Broadcaster.prototype.broadcastBeforeRemoveEvent = function (entity, entityId) {
        var _this = this;
        var subscribers = this.connection
            .subscribers
            .filter(function (subscriber) { return _this.isAllowedSubscribers(subscriber, entity); })
            .filter(function (subscriber) { return !!subscriber.beforeRemove; })
            .map(function (subscriber) { return subscriber.beforeRemove({ entity: entity, entityId: entityId }); });
        var listeners = this.connection.entityListeners
            .filter(function (listener) { return listener.type === EventListenerTypes_1.EventListenerTypes.BEFORE_REMOVE; })
            .filter(function (listener) { return listener.target === entity.constructor; })
            .map(function (entityListener) { return entity[entityListener.propertyName](); });
        return Promise.all(subscribers.concat(listeners)).then(function () { });
    };
    Broadcaster.prototype.broadcastAfterInsertEvent = function (entity) {
        var _this = this;
        var subscribers = this.connection
            .subscribers
            .filter(function (subscriber) { return _this.isAllowedSubscribers(subscriber, entity); })
            .filter(function (subscriber) { return !!subscriber.afterInsert; })
            .map(function (subscriber) { return subscriber.afterInsert({ entity: entity }); });
        var listeners = this.connection.entityListeners
            .filter(function (listener) { return listener.type === EventListenerTypes_1.EventListenerTypes.AFTER_INSERT; })
            .filter(function (listener) { return listener.target === entity.constructor; })
            .map(function (entityListener) { return entity[entityListener.propertyName](); });
        return Promise.all(subscribers.concat(listeners)).then(function () { });
    };
    Broadcaster.prototype.broadcastAfterUpdateEvent = function (entity, updatedColumns) {
        var _this = this;
        var subscribers = this.connection
            .subscribers
            .filter(function (subscriber) { return _this.isAllowedSubscribers(subscriber, entity); })
            .filter(function (subscriber) { return !!subscriber.afterUpdate; })
            .map(function (subscriber) { return subscriber.afterUpdate({ entity: entity, updatedColumns: updatedColumns }); });
        var listeners = this.connection.entityListeners
            .filter(function (listener) { return listener.type === EventListenerTypes_1.EventListenerTypes.AFTER_UPDATE; })
            .filter(function (listener) { return listener.target === entity.constructor; })
            .map(function (entityListener) { return entity[entityListener.propertyName](); });
        return Promise.all(subscribers.concat(listeners)).then(function () { });
    };
    Broadcaster.prototype.broadcastAfterRemoveEvent = function (entity, entityId) {
        var _this = this;
        var subscribers = this.connection
            .subscribers
            .filter(function (subscriber) { return _this.isAllowedSubscribers(subscriber, entity); })
            .filter(function (subscriber) { return !!subscriber.afterRemove; })
            .map(function (subscriber) { return subscriber.afterRemove({ entity: entity, entityId: entityId }); });
        var listeners = this.connection.entityListeners
            .filter(function (listener) { return listener.type === EventListenerTypes_1.EventListenerTypes.AFTER_REMOVE; })
            .filter(function (listener) { return listener.target === entity.constructor; })
            .map(function (entityListener) { return entity[entityListener.propertyName](); });
        return Promise.all(subscribers.concat(listeners)).then(function () { });
    };
    Broadcaster.prototype.broadcastLoadEvents = function (entity) {
        var _this = this;
        var metadata = this.connection.getEntityMetadata(entity.constructor);
        var promises = [];
        metadata
            .relations
            .filter(function (relation) { return entity.hasOwnProperty(relation.propertyName); })
            .map(function (relation) { return entity[relation.propertyName]; })
            .map(function (value) {
            if (value instanceof Array) {
                promises = promises.concat(_this.broadcastLoadEventsForAll(value));
            }
            else {
                promises.push(_this.broadcastLoadEvents(value));
            }
        });
        this.connection
            .subscribers
            .filter(function (subscriber) { return _this.isAllowedSubscribers(subscriber, entity); })
            .filter(function (subscriber) { return !!subscriber.afterLoad; })
            .forEach(function (subscriber) { return promises.push(subscriber.afterLoad(entity)); });
        this.connection
            .entityListeners
            .filter(function (listener) { return listener.type === EventListenerTypes_1.EventListenerTypes.AFTER_LOAD; })
            .filter(function (listener) { return listener.target === entity.constructor; })
            .forEach(function (listener) { return promises.push(entity[listener.propertyName]()); });
        return Promise.all(promises).then(function () { });
    };
    Broadcaster.prototype.broadcastLoadEventsForAll = function (entities) {
        var _this = this;
        return Promise.all(entities.map(function (entity) { return _this.broadcastLoadEvents(entity); })).then(function () { });
    };
    // -------------------------------------------------------------------------
    // Private Methods
    // -------------------------------------------------------------------------
    Broadcaster.prototype.isAllowedSubscribers = function (subscriber, cls) {
        return !subscriber.listenTo ||
            !subscriber.listenTo() ||
            subscriber.listenTo() === Object ||
            subscriber.listenTo() === cls;
    };
    return Broadcaster;
}());
exports.Broadcaster = Broadcaster;
