// typeorm/decorator/relations/ManyToOne.js
"use strict";
var RelationMetadata_1 = require("../../metadata-builder/metadata/RelationMetadata");
var RelationTypes_1 = require("../../metadata-builder/types/RelationTypes");
var MetadataStorage_1 = require("../../metadata-builder/MetadataStorage");
/**
 * Many-to-one relation allows to create type of relation when Entity1 can have single instance of Entity2, but
 * Entity2 can have a multiple instances of Entity1. Entity1 is an owner of the relationship, and storages Entity2 id
 * on its own side.
 */
function ManyToOne(typeFunction, inverseSideOrOptions, options) {
    var inverseSideProperty;
    if (typeof inverseSideOrOptions === "object") {
        options = inverseSideOrOptions;
    }
    else {
        inverseSideProperty = inverseSideOrOptions;
    }
    return function (object, propertyName) {
        if (!options)
            options = {};
        MetadataStorage_1.defaultMetadataStorage.addRelationMetadata(new RelationMetadata_1.RelationMetadata({
            target: object.constructor,
            propertyName: propertyName,
            relationType: RelationTypes_1.RelationTypes.MANY_TO_ONE,
            type: typeFunction,
            inverseSideProperty: inverseSideProperty,
            isOwning: true,
            options: options
        }));
    };
}
exports.ManyToOne = ManyToOne;
