// typeorm/decorator/relations/ManyToMany.js
"use strict";
var RelationMetadata_1 = require("../../metadata-builder/metadata/RelationMetadata");
var RelationTypes_1 = require("../../metadata-builder/types/RelationTypes");
var MetadataStorage_1 = require("../../metadata-builder/MetadataStorage");
/**
 * Many-to-many is a type of relationship when Entity1 can have multiple instances of Entity2, and Entity2 can have
 * multiple instances of Entity1. To achieve it, this type of relation creates a junction table, where it storage
 * entity1 and entity2 ids. This is owner side of the relationship.
 */
function ManyToMany(typeFunction, inverseSideOrOptions, options) {
    var inverseSideProperty;
    if (typeof inverseSideOrOptions === "object") {
        options = inverseSideOrOptions;
    }
    else {
        inverseSideProperty = inverseSideOrOptions;
    }
    return function (object, propertyName) {
        if (!options)
            options = {};
        MetadataStorage_1.defaultMetadataStorage.addRelationMetadata(new RelationMetadata_1.RelationMetadata({
            target: object.constructor,
            propertyName: propertyName,
            relationType: RelationTypes_1.RelationTypes.MANY_TO_MANY,
            type: typeFunction,
            inverseSideProperty: inverseSideProperty,
            isOwning: true,
            options: options
        }));
    };
}
exports.ManyToMany = ManyToMany;
