// typeorm/decorator/relations/OneToMany.js
"use strict";
var RelationMetadata_1 = require("../../metadata-builder/metadata/RelationMetadata");
var RelationTypes_1 = require("../../metadata-builder/types/RelationTypes");
var MetadataStorage_1 = require("../../metadata-builder/MetadataStorage");
/**
 * One-to-many relation allows to create type of relation when Entity2 can have multiple instances of Entity1.
 * Entity1 have only one Entity2. Entity1 is an owner of the relationship, and storages Entity2 id on its own side.
 */
function OneToMany(typeFunction, inverseSideOrOptions, options) {
    var inverseSideProperty;
    if (typeof inverseSideOrOptions === "object") {
        options = inverseSideOrOptions;
    }
    else {
        inverseSideProperty = inverseSideOrOptions;
    }
    return function (object, propertyName) {
        if (!options)
            options = {};
        MetadataStorage_1.defaultMetadataStorage.addRelationMetadata(new RelationMetadata_1.RelationMetadata({
            target: object.constructor,
            propertyName: propertyName,
            relationType: RelationTypes_1.RelationTypes.ONE_TO_MANY,
            type: typeFunction,
            inverseSideProperty: inverseSideProperty,
            isOwning: false,
            options: options
        }));
    };
}
exports.OneToMany = OneToMany;
