// typeorm/decorator/relations/OneToOne.js
"use strict";
var RelationMetadata_1 = require("../../metadata-builder/metadata/RelationMetadata");
var RelationTypes_1 = require("../../metadata-builder/types/RelationTypes");
var MetadataStorage_1 = require("../../metadata-builder/MetadataStorage");
/**
 * One-to-one relation allows to create direct relation between two entities. Entity1 have only one Entity2.
 * Entity1 is an owner of the relationship, and storages Entity1 id on its own side.
 */
function OneToOne(typeFunction, inverseSideOrOptions, options) {
    var inverseSideProperty;
    if (typeof inverseSideOrOptions === "object") {
        options = inverseSideOrOptions;
    }
    else {
        inverseSideProperty = inverseSideOrOptions;
    }
    return function (object, propertyName) {
        if (!options)
            options = {};
        MetadataStorage_1.defaultMetadataStorage.addRelationMetadata(new RelationMetadata_1.RelationMetadata({
            target: object.constructor,
            propertyName: propertyName,
            relationType: RelationTypes_1.RelationTypes.ONE_TO_ONE,
            type: typeFunction,
            inverseSideProperty: inverseSideProperty,
            isOwning: true,
            options: options
        }));
    };
}
exports.OneToOne = OneToOne;
