// typeorm/metadata-builder/EntityMetadataBuilder.js
"use strict";
var TableMetadata_1 = require("./metadata/TableMetadata");
var EntityMetadata_1 = require("./metadata/EntityMetadata");
var ColumnMetadata_1 = require("./metadata/ColumnMetadata");
var ForeignKeyMetadata_1 = require("./metadata/ForeignKeyMetadata");
/**
 * Aggregates all metadata: table, column, relation into one collection grouped by tables for a given set of classes.
 */
var EntityMetadataBuilder = (function () {
    // todo: type in function validation, inverse side function validation
    // -------------------------------------------------------------------------
    // Constructor
    // -------------------------------------------------------------------------
    function EntityMetadataBuilder(metadataStorage, namingStrategy) {
        this.metadataStorage = metadataStorage;
        this.namingStrategy = namingStrategy;
    }
    // -------------------------------------------------------------------------
    // Public Methods
    // -------------------------------------------------------------------------
    /**
     * Builds a complete metadata aggregations for the given entity classes.
     */
    EntityMetadataBuilder.prototype.build = function (entityClasses) {
        var _this = this;
        // filter the metadata only we need - those which are bind to the given table classes
        var tableMetadatas = this.metadataStorage.findTableMetadatasForClasses(entityClasses);
        var abstractTableMetadatas = this.metadataStorage.findAbstractTableMetadatasForClasses(entityClasses);
        var columnMetadatas = this.metadataStorage.findFieldMetadatasForClasses(entityClasses);
        var relationMetadatas = this.metadataStorage.findRelationMetadatasForClasses(entityClasses);
        var indexMetadatas = this.metadataStorage.findIndexMetadatasForClasses(entityClasses);
        var compoundIndexMetadatas = this.metadataStorage.findCompoundIndexMetadatasForClasses(entityClasses);
        var entityMetadatas = tableMetadatas.map(function (tableMetadata) {
            var constructorChecker = function (opm) { return opm.target === tableMetadata.target; };
            var constructorChecker2 = function (opm) { return opm.target === tableMetadata.target; };
            var entityColumns = columnMetadatas.filter(constructorChecker);
            var entityRelations = relationMetadatas.filter(constructorChecker);
            var entityCompoundIndices = compoundIndexMetadatas.filter(constructorChecker2);
            var entityIndices = indexMetadatas.filter(constructorChecker);
            // merge all columns in the abstract table extendings of this table
            abstractTableMetadatas.forEach(function (abstractMetadata) {
                if (!_this.isTableMetadataExtendsAbstractMetadata(tableMetadata, abstractMetadata))
                    return;
                var constructorChecker = function (opm) { return opm.target === abstractMetadata.target; };
                var constructorChecker2 = function (opm) { return opm.target === abstractMetadata.target; };
                var abstractColumns = columnMetadatas.filter(constructorChecker);
                var abstractRelations = entityRelations.filter(constructorChecker);
                var abstractCompoundIndices = entityCompoundIndices.filter(constructorChecker2);
                var abstractIndices = indexMetadatas.filter(constructorChecker);
                var inheritedFields = _this.filterObjectPropertyMetadatasIfNotExist(abstractColumns, entityColumns);
                var inheritedRelations = _this.filterObjectPropertyMetadatasIfNotExist(abstractRelations, entityRelations);
                var inheritedIndices = _this.filterObjectPropertyMetadatasIfNotExist(abstractIndices, entityIndices);
                entityCompoundIndices = entityCompoundIndices.concat(abstractCompoundIndices);
                entityColumns = entityColumns.concat(inheritedFields);
                entityRelations = entityRelations.concat(inheritedRelations);
                entityIndices = entityIndices.concat(inheritedIndices);
            });
            var entityMetadata = new EntityMetadata_1.EntityMetadata(tableMetadata, entityColumns, entityRelations, entityIndices, entityCompoundIndices, []);
            // set naming strategies
            tableMetadata.namingStrategy = _this.namingStrategy;
            entityColumns.forEach(function (column) { return column.namingStrategy = _this.namingStrategy; });
            entityRelations.forEach(function (relation) { return relation.namingStrategy = _this.namingStrategy; });
            return entityMetadata;
        });
        // generate columns and foreign keys for tables with relations
        entityMetadatas.forEach(function (metadata) {
            var foreignKeyRelations = metadata.ownerOneToOneRelations.concat(metadata.manyToOneRelations);
            foreignKeyRelations.map(function (relation) {
                var inverseSideMetadata = entityMetadatas.find(function (metadata) { return metadata.target === relation.type; });
                // find relational columns and if it does not exist - add it
                var relationalColumn = metadata.columns.find(function (column) { return column.name === relation.name; });
                if (!relationalColumn) {
                    var options = {
                        type: inverseSideMetadata.primaryColumn.type,
                        oldColumnName: relation.oldColumnName,
                        nullable: relation.isNullable
                    };
                    relationalColumn = new ColumnMetadata_1.ColumnMetadata({
                        target: metadata.target,
                        propertyName: relation.name,
                        isVirtual: true,
                        options: options
                    });
                    metadata.columns.push(relationalColumn);
                }
                // create and add foreign key
                var foreignKey = new ForeignKeyMetadata_1.ForeignKeyMetadata(metadata.table, [relationalColumn], inverseSideMetadata.table, [inverseSideMetadata.primaryColumn]);
                metadata.foreignKeys.push(foreignKey);
            });
        });
        // set inverse side (related) entity metadatas for all relation metadatas
        entityMetadatas.forEach(function (entityMetadata) {
            entityMetadata.relations.forEach(function (relation) {
                relation.relatedEntityMetadata = entityMetadatas.find(function (m) { return m.target === relation.type; });
            });
        });
        // generate junction tables with its columns and foreign keys
        var junctionEntityMetadatas = [];
        entityMetadatas.forEach(function (metadata) {
            metadata.ownerManyToManyRelations.map(function (relation) {
                var inverseSideMetadata = entityMetadatas.find(function (metadata) { return metadata.target === relation.type; });
                var tableName = metadata.table.name + "_" + relation.name + "_" +
                    inverseSideMetadata.table.name + "_" + inverseSideMetadata.primaryColumn.name;
                var tableMetadata = new TableMetadata_1.TableMetadata(null, tableName, false);
                var column1options = {
                    length: metadata.primaryColumn.length,
                    type: metadata.primaryColumn.type,
                    name: metadata.table.name + "_" + metadata.primaryColumn.name + "_1"
                };
                var column2options = {
                    length: inverseSideMetadata.primaryColumn.length,
                    type: inverseSideMetadata.primaryColumn.type,
                    name: inverseSideMetadata.table.name + "_" + inverseSideMetadata.primaryColumn.name + "_2"
                };
                var columns = [
                    new ColumnMetadata_1.ColumnMetadata({
                        target: null,
                        propertyName: null,
                        options: column1options
                    }),
                    new ColumnMetadata_1.ColumnMetadata({
                        target: null,
                        propertyName: null,
                        options: column2options
                    })
                ];
                var foreignKeys = [
                    new ForeignKeyMetadata_1.ForeignKeyMetadata(tableMetadata, [columns[0]], metadata.table, [metadata.primaryColumn]),
                    new ForeignKeyMetadata_1.ForeignKeyMetadata(tableMetadata, [columns[1]], inverseSideMetadata.table, [inverseSideMetadata.primaryColumn]),
                ];
                var junctionEntityMetadata = new EntityMetadata_1.EntityMetadata(tableMetadata, columns, [], [], [], foreignKeys);
                junctionEntityMetadatas.push(junctionEntityMetadata);
                relation.junctionEntityMetadata = junctionEntityMetadata;
                if (relation.inverseRelation)
                    relation.inverseRelation.junctionEntityMetadata = junctionEntityMetadata;
            });
        });
        return entityMetadatas.concat(junctionEntityMetadatas);
    };
    // -------------------------------------------------------------------------
    // Private Methods
    // -------------------------------------------------------------------------
    EntityMetadataBuilder.prototype.isTableMetadataExtendsAbstractMetadata = function (tableMetadata, abstractMetadata) {
        return tableMetadata.target.prototype instanceof abstractMetadata.target;
    };
    EntityMetadataBuilder.prototype.filterObjectPropertyMetadatasIfNotExist = function (newMetadatas, existsMetadatas) {
        return newMetadatas.filter(function (fieldFromMapped) {
            return existsMetadatas.reduce(function (found, fieldFromDocument) {
                return fieldFromDocument.propertyName === fieldFromMapped.propertyName ? fieldFromDocument : found;
            }, null) === null;
        });
    };
    return EntityMetadataBuilder;
}());
exports.EntityMetadataBuilder = EntityMetadataBuilder;
